### ISSUE TITLE
Support edge based function checkpointing in parallel

### ISSUE THREAD

**Issue by jorgensd** [jorgensd/adios4dolfinx#3]
Need more information about the DOF association to a facet and how they are flipped in dofmap construction. See: https://github.com/FEniCS/dolfinx/issues/2521 for progress in DOLFINx

**Comment by jorgensd** [comment]
New plan, made in collaboration with at @mscroggs and @michalhabera.

Instead of storing a `dofmap` and underlying vector, we could store the function evaluated (in physical space), at it's interpolation points. This means that using `u.interpolate(local_values, local_cells)` would work without having to consider dof permutations.

The steps would be:
Given a function $u\in V$
**Write**
- Get interpolation points on reference element
    - tabulate basis function (at interpolation points)
    - tabulate coordinate map (0th and first order) at interpolation points.
For each cell in V dofmap:
    - apply_dof_transformation to reference basis functions to map them to physical cell
    - compute jacobian, K, detJ
    - push basis values forward
    - Sum up contributions for each point.

These steps would be similar to:
https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/fem/Function.h#L567-L604

**Read**
- Get original input cells as already done in existing code
- Get all values at interpolation points from input cell
- Use https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/fem/interpolate.h#L642-L917 to interpolate data onto given function.

The reason for going down this route is that it seems quite hard to keep track of and applying dof-transformations/inverse transformations, when these has to be done on all the cell_dofs at once, which means a rewrite in communication patterns:
1. Get original cell index as currently done. Also get cell-permutation of partitioned mesh. Tuple (`oci, cp`)
2. Send (`oci, cp`) from cell owner (`c_o`) to input owner (`i_o`). For all dofs in `input_dofmap.cell_dofs(oci-local_range[0])` figure out where to fetch the `u_input.x.array` entries (`a_o`).
3. Get the `u_input.x.array` from (`a_o`) to (`i_o`).
4. Compute inverse transformation of input array on `i_o` using `cell_permutation` from file. 
5. Compute transformation using `cp` from `c_o` on `i_o`.
6. Send all cell values back to `c_o`.
7. Scatter forward (will overwrite several values already retrieved).

**Comment by IgorBaratta** [comment]
I believe that the first option would result in a simpler solution at the expense of more data being stored:

```python3
mesh = dolfinx.mesh.create_unit_cube(MPI.COMM_WORLD, 10, 10 ,10, cell_type=dolfinx.mesh.CellType.hexahedron)
V = dolfinx.fem.FunctionSpace(mesh, ("NCE", 1))
points = V.element.interpolation_points()
ncells = mesh.topology.index_map(3).size_global
print(3 * ncells * points.shape[0])
print(V.dofmap.index_map.size_global)
```
Storing data at interpolation points would lead to a 10x increase in data to be stored:
```bash
36000
3630
```
